### PR TITLE
Add better error message for empty and null -UFormat arg

### LINKED ISSUE #5047: Get-Date -UFormat '' throws an unhelpful exception for an empty string argument
An empty string argument in `Get-Date -UFormat ''` throws `IndexOutOfRangeException`, it would be more helpful if the error said what was wrong and directed attention to the UFormat argument. "Missing or Invalid UFormat String" or words to that effect.

Steps to reproduce
------------------

```powershell
PS /test> get-date -uformat "$accidentallyEmptyFormatStringVariable"
get-date : Index was outside the bounds of the array.
At line:1 char:1
+ get-date -uformat ''
+ ~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Get-Date], IndexOutOfRangeException
    + FullyQualifiedErrorId : System.IndexOutOfRangeException,Microsoft.PowerShell.Commands.GetDateCommand

```

Expected behavior
-----------------

```none
A message directing attention to the -UFormat string parameter being empty.
```

Actual behavior
---------------

```none
A message about some internal problem.
```

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      6.0.0-beta.8
PSEdition                      Core
GitCommitId                    v6.0.0-beta.8
OS                             Linux 2.6.32-042stab120.6 #1 SMP Thu Oct 27 16:59:03 MSK 2016
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

```

### PR DESCRIPTION
**Pull request by djweber**
Previously, when passing an empty format string to -UFormat, a somewhat unhelpful
error message would appear. This commit fixes #5047 by adding a more descriptive error to better
guide the user

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/resources/GetDateStrings.resx
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Utility/resources/GetDateStrings.resx]: Should have sentence end with period.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1]: Perhaps have two `-TestCases` where one is an empty string and the other is just whitespace

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]
Thanks for the contribution!  Just a few minor comments

**Comment by DdWr** [comment]
> Thanks for the contribution! Just a few minor comments
> Show outdated src/Microsoft.PowerShell.Commands.Utility/resources/GetDateStrings.resx
> Show outdated test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1

Thanks for the feedback! I made the changes you requested.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: What about `null`? And I wonder why not use standard `[ValidateNotNullOrEmpty()]`?

Comment (DdWr) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: It seems that `[ValidateNotNullOrEmpty()]` does not take whitespace-only strings into account, so it looks like that has to be checked manually. As for `null`, it looks like the way the current code is written, if `$null` is passed to `-uformat`, it will just fall through the if-else completely and just print the date with a default format. Do we want to keep this behavior, or do we want to use [ValidateNotNull()]?

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: " " - is allowed format string.
Disabling Null is breaking change - @SteveL-MSFT What is your thoughts?

**Comment by DdWr** [comment]
> " " - is allowed format string.
> Disabling Null is breaking change - @SteveL-MSFT What is your thoughts?

Just to clarify, I haven't changed the behavior of `$null` being passed in. I did not know that `" "` was a valid format string -- I assumed that this was also considered empty.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs
Comment (DdWr) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: Thanks for the feedback. Just to clarify, I haven't changed the behavior of $null being passed in. I did not know that " " was a valid format string -- I assumed that this was also considered empty.

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: Since `$null` is the same as if `-UFormat` wasn't provided, we probably don't want to break that behavior as someone may be relying on it.  `-UFormat` enables custom formatting, so any valid string should work.  I suppose it's not different if they provided `" "` or `"apple"` which should assume they know what they are doing.  So it seems that in this case, the only fix is on [this line](https://github.com/DdWr/PowerShell/blob/d7fc318c63e9192148fb100070b13deb9ea0266d/src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs#L349) if UFormat.Length == 0, just return.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: We can use `ValidateLength`

Comment (DdWr) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: @SteveL-MSFT @iSazonov Thank you for the responses. Based off this discussion, can we clarify the desired behavior? Based off of what has been said, this is the behavior it seems that we want here:

**Current**
* If `$null` is passed in, fall through and just print the date
* If `'  '`, `apple`, or any unconventional date format of length > 0 is passed in, keep the current behavior and show the formatted result.

**Proposed addition**
* Add a length check to [this line](https://github.com/DdWr/PowerShell/blob/d7fc318c63e9192148fb100070b13deb9ea0266d/src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs#L349), which will cause it to match the above behavior and format the date to an empty string.
* Use `ValidateLength`, which will throw an error if an empty string is passed in.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: I believe ` ValidateNotNullOrEmpty ` would best solution if @SteveL-MSFT'd agreed. This breaking change seems incredible.

Comment (DdWr) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: Updated PR with proposed changes.

**Comment by iSazonov** [comment]
CI MacOs temporary failed - I restart it.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: @SteveL-MSFT I think the uformat is static constant in most scenarios. If someone dynamically forms it, it's not likely that it will leave it uninitialized (null). So I believe this is a breaking change  only formally and we can accept it.

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: I agree that most likely if $null gets passed, it's probably a problem in someone's script.  Let me get a quick ack from @PowerShell/powershell-committee before we take this breaking change.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and agree that using the ValidateNotNullOrEmpty attribute is the right fix and having it error with $null is the right behavior although breaking it seems low risk of impact.

**Comment by iSazonov** [comment]
@DdWr Please correct your commit - it contains alien code.

**Comment by DdWr** [comment]
@iSazonov Fixed

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1]: Please add test for $null.
Use `TestCases`.

Comment (DdWr) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1]: @iSazonov PR updated

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1]: I would suggest putting single quotes around it to make it more readable in the test output:

```none
It "Passing '<name>' to ...
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1]: Please change `"null"` to `"$null"` (use backtick to escape the dollar sign).  Also add some whitespace to align `value` with test case below to make it easier to read:

```powershell
@{ name = "`$null"      ; value = $null }
@{ name = "empty string"; value = "" }
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1]: Suggesting removing `an` and adding whitespace after the empty string (see above)

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]
Thanks for the contribution.  Just a few minor comments.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1
Comment (DdWr) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1]: @SteveL-MSFT Done

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1
Comment (DdWr) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1]: @SteveL-MSFT Done. Thank you for the feedback -- it will be helpful moving forward.

**Review by SteveL-MSFT — APPROVED** [review]
LGTM!

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@DdWr Please don't rebase your commits and keep a _history_. Sometimes we have to rebase to resolve merge conflitcs. Ask maintainers If you are not sure.

**Pull request merged**
